bookkeeper-server/src/main/java/org/apache/bookkeeper/client/LedgerHandle.java

```diff
@@ -1898,302 +1788,81 @@ void handleBookieFailure(final Map<Integer, BookieSocketAddress> failedBookies)
-        }
+        new MetadataUpdateLoop(
+                clientCtx.getLedgerManager(), getId(),
+                this::getLedgerMetadata,
+                (metadata) -> {
+                    List<BookieSocketAddress> lastEnsemble = metadata.getLastEnsembleValue();
+                    boolean failedBookieInEnsemble = failedBookies.entrySet().stream()
+                        .anyMatch((e) -> lastEnsemble.get(e.getKey()).equals(e.getValue()));
+                    return !metadata.isClosed() && !metadata.isInRecovery() && failedBookieInEnsemble;
+                },
+                (metadata) -> {
+                    List<BookieSocketAddress> newEnsemble = EnsembleUtils.replaceBookiesInEnsemble(
+                            clientCtx.getBookieWatcher(), getId(), metadata, getCurrentEnsemble(), failedBookies);
+                    Long lastEnsembleKey = metadata.getLastEnsembleKey();
+                    LedgerMetadataBuilder builder = LedgerMetadataBuilder.from(metadata);
+                    long newEnsembleStartEntry = getLastAddConfirmed() + 1;
+                    checkState(lastEnsembleKey <= newEnsembleStartEntry,
+                               "New ensemble must either replace the last ensemble, or add a new one");
+                    if (lastEnsembleKey.equals(newEnsembleStartEntry)) {
+                        return builder.replaceEnsembleEntry(newEnsembleStartEntry, newEnsemble).build();
+                    } else {
+                        return builder.newEnsembleEntry(newEnsembleStartEntry, newEnsemble).build();
+                    }
+                },
+                this::setLedgerMetadata)
+            .run().whenCompleteAsync((metadata, ex) -> {
+                    if (ex != null) {
+                        handleUnrecoverableErrorDuringAdd(BKException.getExceptionCode(ex, WriteException));
+                    } else if (metadata.isClosed()) {
+                        handleUnrecoverableErrorDuringAdd(BKException.Code.LedgerClosedException);
+                    } else if (metadata.isInRecovery()) {
+                        handleUnrecoverableErrorDuringAdd(BKException.Code.LedgerFencedException);
+                    } else {
+                        synchronized (metadataLock) {
+                            if (!delayedWriteFailedBookies.isEmpty()) {
+                                Map<Integer, BookieSocketAddress> toReplace = new HashMap<>(delayedWriteFailedBookies);
+                                delayedWriteFailedBookies.clear();
+
+                                ensembleChangeLoop(origEnsemble, toReplace);
+                            } else {
+                                List<BookieSocketAddress> newEnsemble = getCurrentEnsemble();
+                                Set<Integer> replaced = EnsembleUtils.diffEnsemble(origEnsemble, newEnsemble);
+                                unsetSuccessAndSendWriteRequest(newEnsemble, replaced);
+                                changingEnsemble = false;
```

Comment:
I think we have to call sendAddSuccessCallbacks after calling unsetSucccessAndSendWriteRequest.  It's possible that every pendingAddOp had a bookie in the write set swapped, but nevertheless all have ackQuorum satisfied.  In that case, PendingAddOp.unsetSuccessAndSendWriteRequest will not call sendAddSuccessCallbacks(), but we will still have writes ready to go.  In fact, I think the call in PendingAddOp.unsetSuccessAndSendWriteRequest is entirely unnecessary and should be removed in favor of a single call here.  As a side effect, you can call it outside of the metadataLock avoiding that problem as well.

Reply:
I agree. Will change this.

Reply:
> I agree. Will change this.

PendingAddOp.unsetSuccessAndSendWriteRequest() will do a sendWriteRequest() even if the AQ satisfied. When that write request response comes back irrespective of a pass/fail it will do sendAddSuccessCallbacks().

Reply:
@jvrao you're right. This stuff jumps far too much between LedgerHandle and PendingAddOp. I'll leave this for now.